File: maven-projects/messaging-client/src/main/java/net/dushin/lethe/messaging/client/crypto/Decryptor.java
/**
 * Copyright (c) dushin.net
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *     * Redistributions of source code must retain the above copyright
 *       notice, this list of conditions and the following disclaimer.
 *     * Redistributions in binary form must reproduce the above copyright
 *       notice, this list of conditions and the following disclaimer in the
 *       documentation and/or other materials provided with the distribution.
 *     * Neither the name of dushin.net nor the
 *       names of its contributors may be used to endorse or promote products
 *       derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY dushin.net ``AS IS'' AND ANY
 * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL dushin.net BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
package net.dushin.lethe.messaging.client.crypto;

import net.dushin.lethe.messaging.interfaces.EncryptedKey;
import net.dushin.lethe.messaging.interfaces.EncryptedKeyList;
import net.dushin.lethe.messaging.interfaces.EncryptedMessage;

/**
 * This class is used to decrypt encrypted messages.  It must be initialized
 * with a private key, which must correspond with the public key used to
 * encrypt the message.
 */
public class Decryptor extends CryptorBase {

    /**
     * @param       key
     *              The private key used to decrypt messages
     */
    public
    Decryptor(
        final java.security.PrivateKey key
    ) {
        super(
            "RSA/ECB/NoPadding",
            javax.crypto.Cipher.DECRYPT_MODE,
            key
        );
    }
    
    /**
     * Decrypt an EncryptedMessage.
     *
     * @param       encrypted
     *              The EncryptedMessage to decrypt.
     *
     * @return      the decrypted message.  Typically, this object will
     *              be an instance of a PlaintextMessage or a SignedMessage,
     *              depending on whether the sender has signed the encrypted
     *              message.
     */
    public Object
    decrypt(final EncryptedMessage encrypted) {
        try {
            final javax.crypto.SecretKey decryptedKey = 
                findDecryptedKey(
                    encrypted.getRecipients(), encrypted.getAlgorithm()
                );
            final byte[] decryptedData = decryptData(decryptedKey, encrypted.getEncryptedData());
            return deserialize(
                EncryptedMessage.class.getPackage(),
                decryptedData
            );
        } catch (final Exception e) {
            throw new RuntimeException("Error decrypting", e);
        }
    }
    
    //
    // internal operations
    //
    
    /**
     * @return      the decrypted secret key, from a list of encrypted keys.
     *              This operation will return a non-null value if there is
     *              an encrypted key in the specified list that was encrypted
     *              using the public key corresponding to the private key
     *              with which this Decryptor was initialized.  Otherwise,
     *              the operation will raise an exception.
     */
    private javax.crypto.SecretKey
    findDecryptedKey(
        final EncryptedKeyList encryptedKeys,
        final String algorithm
    ) {
        for (EncryptedKey encryptedKey : encryptedKeys.getItem()) {
            try {
                return decryptKey(encryptedKey, algorithm);
            } catch (final Exception e) {
                continue;
            }
        }
        throw new RuntimeException("Key not found");
    }
    
    /**
     * @return      the decrypted secret key, from the specified EncryptedKey.
     *              This operation will raise an exception if decryption fails.
     */
    private javax.crypto.SecretKey
    decryptKey(
        final EncryptedKey encryptedKey,
        final String algorithm
    ) {
        try {
            final byte[] encryptedKeyData = encryptedKey.getData();
            final byte[] decryptedKey = this.cipher.doFinal(encryptedKeyData);
            Logger.logBuffer(
                "Encrypted key prior to decryption:",
                encryptedKeyData
            );
            Logger.logBuffer(
                "Decrypted key:",
                decryptedKey
            );
            //
            // TODO something is awry here -- I need to ignore the first
            // 32 bytes of the decrypted data -- they are all 0 for some
            // reason I need to look into
            //
            final javax.crypto.spec.SecretKeySpec spec = 
                new javax.crypto.spec.SecretKeySpec(
                    decryptedKey, 
                    32, 32, 
                    algorithm
                );
            return spec;
        } catch (final Exception e) {
            throw new RuntimeException("Error decrypting", e);
        }
    }
    
    /**
     * @return      the result of decrypting the specified data with the
     *              specified secret key.
     */
    private static byte[]
    decryptData(
        final javax.crypto.SecretKey key,
        final byte[] data
    ) {
        try {
            final SymmetricDecryptor decryptor = new SymmetricDecryptor(key);
            return decryptor.decrypt(data);
        } catch (final Exception e) {
            throw new RuntimeException("Error decrypting", e);
        }
    }
}

File: dashcode/Lethe.dcproj-ci/project/crypto/decryptor.js
//
// Copyright (c) dushin.net
// All rights reserved.
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are met:
//     * Redistributions of source code must retain the above copyright
//       notice, this list of conditions and the following disclaimer.
//     * Redistributions in binary form must reproduce the above copyright
//       notice, this list of conditions and the following disclaimer in the
//       documentation and/or other materials provided with the distribution.
//     * Neither the name of dushin.net nor the
//       names of its contributors may be used to endorse or promote products
//       derived from this software without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY dushin.net ``AS IS'' AND ANY
// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
// WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
// DISCLAIMED. IN NO EVENT SHALL dushin.net BE LIABLE FOR ANY
// DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
// (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
// LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
// ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
//

// requires function imprt
// requires function net_dushin_exception.ExceptionFactory.createException
// requires function net_dushin_exception.ExceptionFactory.createIllegalArgumentException
__uses("binary.js");
__uses("Cipher.js");

/**
 * An DecryptorFactory is used to create an decryptor object,
 * which in turn is used to edecrypt messages.
 *
 * A decryptor is created using the DecryptorFactory#createDecryptor
 * method, which is passed a paramter specification.
 *
 * Typical usage:
 *
 * var decryptor = DecryptorFactory.create({privateKey: ...});
 * var message = encryptor.decrypt(encryptedMessage)
 */
net_dushin_crypto.DecryptorFactory = {
    
    /**
     * The Cipher type
     */
    Cipher: __import(this, "titaniumcore.crypto.Cipher"),
        
    /**
     * The JSON-RPC object (for marshalling)
     */
    jsonrpc: JSOlait.imprt("jsonrpc"),
    
    /**
     * object createDecryptor(spec)
     *
     * @param       spec
     *              The constructor specification.  This parameter may contain
     *              the following elements:
     *
     *                  * privateKey:       the private key used to decrypt messages.
     *                                      This parameter is required
     *
     * @return      an decryptor object, implements the decrypt method
     *
     * @exception   IllegalArgument, if the private key is not provided
     */
    createDecryptor: function(spec) {
        //
        // Check the constructor parameters for a private key, to be used to
        // perform decryption.
        //
        if (!spec.privateKey) {
            throw net_dushin_foundation.ExceptionFactory.createIllegalArgumentException( 
                {message: "Missing privateKey parameter"}
            );
        }
        
        /**
         * the root scope
         */
        var root = this;
        
        /**
         * the private key
         */
        var rsa = net_dushin_crypto.KeyUtil.parseEncodedPrivateKey(spec.privateKey);
        
        //
        // Create and return the encryptor
        //
        return {
            /**
             * object decrypt(object)
             *
             * @param       object
             *              The object to decrypt.
             *
             * @return      the encrypted object, conforming to the definition at XXX
             */
            decrypt: function(object) {
                if (object.type !== "encrypted") {
                    throw net_dushin_foundation.ExceptionFactory.createException(
                        {
                            message: "Message is not a signed object message."
                        }
                    );
                }
                var cipher = root.Cipher.create(object.algorithm, "DECRYPT", object.mode, object.padding);
                var symmetricKey;
                if (object.encryptedKey) {
                    symmetricKey = rsa.privateDecrypt(base64_decode(object.encryptedKey));
                } else if (object.encryptedKeys) {
                    symmetricKey = net_dushin_foundation.Lists.mapFirst(
                        function(encryptedKey) {
                            try {
                                return rsa.privateDecrypt(base64_decode(encryptedKey));
                            } catch (e) {
                                return false;
                            }
                        },
                        object.encryptedKeys
                    );
                    if (!symmetricKey) {
                        throw "Unable to decrypt any of the encrypted symmetric keys";
                    }
                } else {
                    throw "Encrypted message must contain at most one of encryptedKey or encryptedKeys";
                }
                var plaintext = utf82str(cipher.execute(symmetricKey.slice(0, 16), base64_decode(object.ciphertext)));
                return root.jsonrpc.unmarshall(plaintext);
            }
        };
    }
};
